### PR TITLE
Figure out the type of modified residues 

### PR DESCRIPTION
**Pull request by LesikDee**
Added an algorithm for finding out the type of modified residues (nucleic or amino) by other residues in one chain

## Description

Residues in a chain are either amino acid either nucleic acid (and water). There might be some modified/mutated residues, which type could not be determined by their name (nucleic or amino); In this case firstly program finds out the chain type (by well-known residues) and then propagates it to modified/mutated residues.
Also, water residues now handled as nucleic (before as amino).

This PR resolved  `781` issue, besides testing correct program behavior on declared files (1VRO, 1K1R)  it was found 2 another modified DNA (2MMF, 2H0N), which before these changes miew couldn't handle. 
 
## Type of changes

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read [CONTRIBUTING](https://github.com/epam/miew/blob/master/CONTRIBUTING.md) and [CODE_OF_CONDUCT](https://github.com/epam/miew/blob/master/CODE_OF_CONDUCT.md) guides.
- [x] I have followed the code style of this project.
- [x] I have run `yarn run ci`: lint and tests pass locally with my changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation / The changes do not need docs update.

### CODE REVIEW

File: src/chem/Chain.js [hunk 1]
```diff
@@ -2,6 +2,15 @@ import * as THREE from 'three';
 import Residue from './Residue';
 import ResidueType from './ResidueType';
 
+/**
+ * Residues in chain are either amino acid either nucleic acid (and water)
+ * There might be some modified/mutated residues, which could not be definite by their name (nucleic or amino); In this
+ * case firstly program definites the chain type (by well-known residues) and then definites modified/mutated residues
+ */
+const CHAIN_TYPE = {
+  INDEFINITE: 0, NUCLEIC: 1, AMINO: 2,
```
Comment (paulsmirnov) [inline src/chem/Chain.js]: I'd prefer UNKNOWN, UNDEFINED, or even GENERIC to 'indefinite'

File: src/chem/Chain.js [hunk 2]
```diff
@@ -2,6 +2,15 @@ import * as THREE from 'three';
 import Residue from './Residue';
 import ResidueType from './ResidueType';
 
+/**
+ * Residues in chain are either amino acid either nucleic acid (and water)
+ * There might be some modified/mutated residues, which could not be definite by their name (nucleic or amino); In this
```
Comment (paulsmirnov) [inline src/chem/Chain.js]:
```suggestion
 * There might be some modified/mutated residues, which type could not be determined by their name (nucleic or amino); In this
```

File: src/chem/Chain.js [hunk 3]
```diff
@@ -36,6 +46,38 @@ class Chain {
     return this._residues;
   }
 
+  getType() {
+    return this._type;
+  }
```
Comment (paulsmirnov) [inline src/chem/Chain.js]: Let's avoid unnecessary getter/setters for public properties in the new code (they are all public, actually). Please remove this method and the underscore from the `.type`.

File: src/chem/Chain.js [hunk 4]
```diff
@@ -2,6 +2,15 @@ import * as THREE from 'three';
 import Residue from './Residue';
 import ResidueType from './ResidueType';
 
+/**
+ * Residues in chain are either amino acid either nucleic acid (and water)
+ * There might be some modified/mutated residues, which could not be definite by their name (nucleic or amino); In this
+ * case firstly program definites the chain type (by well-known residues) and then definites modified/mutated residues
+ */
+const CHAIN_TYPE = {
```
Comment (paulsmirnov) [inline src/chem/Chain.js]: By the way, why is it all caps?
```suggestion
const ChainType = {
```

File: src/chem/Chain.js [hunk 5]
```diff
@@ -36,6 +46,38 @@ class Chain {
...
+    const residues = this._residues;
+
+    for (let i = 0, n = residues.length; i < n; ++i) {
+      if ((residues[i]._type.flags & ResidueType.Flags.NUCLEIC) !== 0) {
+        this._type = CHAIN_TYPE.NUCLEIC;
+        break;
+      } else if ((residues[i]._type.flags & ResidueType.Flags.PROTEIN) !== 0) {
+        this._type = CHAIN_TYPE.AMINO;
+        break;
+      }
```
Comment (paulsmirnov) [inline src/chem/Chain.js]: Should we rely on the value filled in the constructor or reset the type to UNKNOWN/GENERIC just before the loop?

File: src/chem/Chain.js [hunk 6]
```diff
@@ -36,6 +46,38 @@ class Chain {
...
+    }
+
+    for (let i = 0, n = residues.length; i < n; ++i) {
+      if (residues[i]._type.flags === ResidueType.Flags.UNDEFINED) {
```
Comment (paulsmirnov) [inline src/chem/Chain.js]: 1. Please cache the repeating `residues[i]._type` as `residueType` for the sake of speed and size.
2. `ResidueType.Flags` are **bit flags** for various combinations! You cannot just compare or assign them (as below).
3. I don't think `ResidueType.Flags.UNDEFINED` will help.

Another my concern is that you are assigning the type of all residues in a chain, even ligands. Please check 4NRE.pdb for FE2, C8E, SO4 residues. They are neither PROTEIN nor NUCLEIC, and other algorithms may fail after such an assignment.

We need a different solution, e.g.
- do not guess residue types now,
- introduce `PROTEIN_CHAIN` and `NUCLEIC_CHAIN` if you really must.

Let's discuss.

File: src/chem/Residue.js [hunk 7]
```diff
@@ -197,7 +197,8 @@ class Residue {
 
     const lp = getAtomPosition(this._leadAtom);
     const currLeadPos = new THREE.Vector3(lp.x, lp.y, lp.z);
-    if ((this._type.flags & ResidueType.Flags.NUCLEIC) !== 0) {
+    if (((this._type.flags & ResidueType.Flags.NUCLEIC) !== 0
+    || (this._type.flags & ResidueType.Flags.WATER) !== 0)) {
```
Comment (paulsmirnov) [inline src/chem/Residue.js]: Shouldn't there be an indentation? And what about other residue types/classes (e.g. added in the future)? Shouldn't we invert the condition or change it in some other way to make generic?

File: src/chem/ResidueType.js [hunk 8]
```diff
@@ -15,13 +15,54 @@ class ResidueType {
...
+    PYRIMIDINE: 0x0400,
+    /** DNA */
+    DNA: 0x0800,
+    /** RNA */
+    RNA: 0x1000,
+
+    /** Water */
+    WATER: 0x10000,
+  };
+
```
Comment (paulsmirnov) [inline src/chem/ResidueType.js]: Why move it? Is there a way to avoid such a big diff?

File: src/chem/Chain.js (hunk 6 above)
Comment (LesikDee) [reply src/chem/Chain.js]: The idea about ResidueType.Flags.UNDEFINED is that before my changes if the residue is neither nucleic neither protein neither water then its residues[i]._type is equal to 0x0000 - for me that is the magic number, I decided to call it. 

The algorithm handles two principle different cases: if the residue is nucleic or not. So to all residues that aren't nucleic, I assign type "Amino". Yes, it seems, that there are residues that are not either nucleic either amino, but the principle difference: are they nucleic or not.
So, I see the decision to use enum like NUCLEIC, NOT_NUCLEIC.  
What do you think about this?
Interest fact: water residues need to be handled as NUCLEIC.

File: src/chem/Chain.js (hunk 1 above)
Comment (LesikDee) [reply src/chem/Chain.js]: Yes, I think, me too. Will correct!

File: src/chem/Chain.js (hunk 3 above)
Comment (LesikDee) [reply src/chem/Chain.js]: Got it

File: src/chem/Chain.js (hunk 4 above)
Comment (LesikDee) [reply src/chem/Chain.js]: Saw on enum in "gfxutils", will correct that.

File: src/chem/Residue.js (hunk 7 above)
Comment (LesikDee) [reply src/chem/Residue.js]: I will think about it, thanks

**Comment by LesikDee** [comment]
Thank you for your comments! I will correct the drawbacks and write some tests, that would prove my fix working as I want to.

File: src/chem/Chain.js (hunk 6 above)
Comment (paulsmirnov) [reply src/chem/Chain.js]: Finally, we decided to choose a slightly different approach.

File: src/chem/ResidueType.js (hunk 8 above)
Comment (paulsmirnov) [reply src/chem/ResidueType.js]: Ok, we solved the issue

**Pull request merged**